File: base/hourai/bot/extensions/auto/message_filter.py
import asyncio
import discord
import re
import logging
from discord.ext import commands
from hourai import utils
from hourai.bot import cogs
from hourai.db import proto
from hourai import config as hourai_config
from hourai.utils import embed as embed_utils
from hourai.utils import invite, mentions, format


def generalize_filter(filter_value):
    filter_value = re.escape(filter_value)

    def _generalize_character(char):
        return char + '+' if char.isalnum() else char
    return ''.join(_generalize_character(char) for char in filter_value)


def make_slur_filter():
    slurs = hourai_config.load_list(hourai_config.get_config(),
                                    'message_filter_slurs')
    components = [generalize_filter(s) for s in slurs]
    regex = f"({'|'.join(components)})"
    logging.info(f"Slur Filter: {regex}")
    return re.compile(regex)


SLUR_FILTER = make_slur_filter()


class MessageFilter(cogs.BaseCog):

    def __init__(self, bot):
        super().__init__()
        self.bot = bot

    @commands.Cog.listener()
    async def on_message(self, message):
        await self.check_message(message)

    @commands.Cog.listener()
    async def on_raw_messsage_edit(self, payload):
        await self.check_message(payload)

    async def check_message(self, message):
        config = await self.get_mod_config(message)
        if config is None:
            return

        for rule in config.message_filter.rules:
            reasons = await self.get_rule_reason(message, rule.criteria)
            if reasons:
                await self.apply_rule(rule, message, reasons)

    async def get_mod_config(self, message):
        proxy = self.bot.get_guild_proxy(message.guild)
        if proxy is None:
            return None

        if isinstance(message, discord.RawMessageUpdateEvent):
            try:
                channel = self.bot.get_channel(message.channel_id)
                message = await channel.fetch_message(message.message_id)
            except (AttributeError, discord.NotFound, discord.Forbidden):
                return

        has_filter = proxy.config.moderation.HasField('message_filter')
        is_bot_user = message.author == self.bot.user
        in_modlog = proxy.config.logging.modlog_channel_id == message.channel.id

        if has_filter and not (in_modlog and is_bot_user):
            return mod_config
        return None

    async def apply_rule(self, rule, message, reasons):
        tasks = []
        action_taken = ""
        mention_mod = rule.notify_moderator
        reasons_block = f"\n```\n{format.vertical_list(reasons)}\n```"
        guild = self.bot.get_guild_proxy(message.guild)

        if rule.notify_moderator:
            action_taken = "Message filter found notable message:"
        if rule.delete_message:
            permissions = message.channel.permissions_for(guild.me)
            if permissions.manage_messages:
                if rule.notify_moderator:
                    action_taken = "Message filter deleted message:"

                dm = (f"[{guild.name}] Your message was deleted for "
                      f"the following reasons: {reasons_block}")

                async def delete():
                    await message.delete()
                    if not message.author.bot and \
                       message.author != self.bot.user:
                        await message.author.send(dm)
                tasks.append(delete())
            else:
                mention_mod = True
                action_taken = (f"Attempted to delete, but don't have "
                                f"`Manage Messages` in "
                                f"{message.channel.mention}.")
        if rule.additional_actions:
            actions = []
            for action_template in rule.additional_actions:
                action = proto.Action()
                action.CopyFrom(action_template)
                action.guild_id = guild.id
                action.user_id = message.author.id
                if not action.HasField('reason'):
                    action.reason = f"Triggered message filter: '{rule.name}'"
                actions.append(action)
            tasks.append(
                    self.bot.action_manager.sequentially_execute(actions))

        if mention_mod or action_taken:
            text = action_taken + reasons_block
            if mention_mod:
                _, mention_text = utils.mention_random_online_mod(guild)
                text = mention_text + " " + text
            embed = embed_utils.message_to_embed(message)
            tasks.append(guild.get_modlog().send(content=text, embed=embed))

        try:
            await asyncio.gather(*tasks)
        except discord.Forbidden:
            pass

    async def get_rule_reason(self, message, criteria):
        reasons = []
        for regex in criteria.matches:
            if re.search(regex, message.content):
                reasons.append("Message contains banned word or phrase.")

        if criteria.includes_slurs:
            for word in message.content.split():
                if SLUR_FILTER.match(word):
                    reasons.append(
                            f"Message contains recognized racial slur: {word}")
                    break

        if criteria.includes_invite_links:
            if any(invite.get_discord_invite_codes(message.content)):
                reasons.append("Message contains Discord invite link.")

        reasons += list(self.get_mention_reason(message, criteria.mentions))
        reasons += list(self.get_embed_reason(message, criteria.embeds))

        exclude_criteria = (
            # Exclude the owner of the bot and the owner of the server.
            await self.bot.is_owner(message.author),
            message.guild.owner_id == message.author.id,
            # Exclude moderators and bots if configured.
            criteria.exclude_moderators and utils.is_moderator(message.author),
            criteria.exclude_bots and message.author.bot,
            # Exclude specific channels when configured.
            message.channel.id in criteria.excluded_channels,
        )
        if any(exclude_criteria):
            reasons.clear()
        return reasons

    def get_mention_reason(self, message, criteria):
        def check_counts(name, limits, msg_mentions):
            unique_mentions = set(msg_mentions)
            if limits.HasField('maximum_total') and \
               len(msg_mentions) > limits.maximum_total:
                yield (f"Total {name} more than the server limit "
                       f"({limits.maximum_total}).")
            if limits.HasField('maximum_unique') and \
               len(unique_mentions) > limits.maximum_unique:
                yield (f"Unique {name} more than the server limit "
                       f"({limits.maximum_unique}).")

        users = mentions.get_user_mention_ids(message.content)
        roles = mentions.get_role_mention_ids(message.content)

        user_mentions = ["u{id}" for id in users]
        role_mentions = ["r{id}" for id in roles]
        all_mentions = user_mentions + role_mentions

        yield from check_counts("user mentions", criteria.role_mention,
                                user_mentions)
        yield from check_counts("role mentions", criteria.role_mention,
                                role_mentions)
        yield from check_counts("mentions", criteria.any_mention, all_mentions)

    def get_embed_reason(self, message, criteria):
        embed_urls = {e.url for e in message.embeds}
        attachment_urls = {a.url for a in message.attachments}
        unique_embeds = embed_urls | attachment_urls
        if criteria.HasField('max_embed_count') and \
           len(unique_embeds) > criteria.max_embed_count:
            yield (f"Message has {len(unique_embeds)} embeds or attachments. "
                   f"More than the server maximum of "
                   f"{criteria.max_embed_count}.")

File: logger/hourai-logger/src/message_filter.rs
use crate::message_logging;
use anyhow::Result;
use hourai::proto::guild_configs::*;
use hourai::{
    models::{
        id::{ChannelId, RoleId},
        message::MessageLike,
        user::UserLike,
        Snowflake,
    },
    util::mentions,
};
use hourai_redis::{CachedMessage, GuildConfig};
use hourai_sql::Member;
use hourai_storage::actions::ActionExecutor;
use regex::{Regex, RegexSet};
use std::collections::HashSet;

lazy_static! {
    static ref SLUR_REGEX: RegexSet = generalize_filters(SLURS);
    static ref DISCORD_INVITE_REGEX: Regex = Regex::new("discord.gg/([a-zA-Z0-9]+)").unwrap();
}

const SLURS: &[&str] = &[
    "nigger", "nigga", "tarskin", "tranny", "trannie", "redskin", "faggot", "chink", "kike",
    "dyke", "gook", "wigger",
];

pub async fn check_message(executor: &ActionExecutor, message: &impl MessageLike) -> Result<bool> {
    let guild_id = if let Some(guild_id) = message.guild_id() {
        guild_id
    } else {
        return Ok(false);
    };
    let mut storage = executor.storage().clone();
    let config: ModerationConfig = GuildConfig::fetch_or_default(guild_id, &mut storage).await?;

    let mut redis = storage.redis().clone();
    let member = Member::fetch(guild_id, message.author().id())
        .fetch_one(executor.storage())
        .await;
    let moderator = if let Ok(member) = member {
        hourai_storage::is_moderator(guild_id, member.role_ids(), &mut redis).await?
    } else {
        false
    };

    for rule in config.get_message_filter().get_rules() {
        let reasons = get_filter_reasons(moderator, message, rule.get_criteria()).await?;
        if !reasons.is_empty() {
            apply_rule(message, rule.clone(), reasons, executor).await?;
            return Ok(rule.get_delete_message());
        }
    }

    Ok(false)
}

fn generalize_filters(filters: &[&str]) -> RegexSet {
    let generalized = filters
        .iter()
        .map(|filter| generalize_filter(*filter))
        .collect::<Vec<String>>();
    RegexSet::new(&generalized).unwrap()
}

fn generalize_filter(filter: &str) -> String {
    regex::escape(filter)
        .chars()
        .map(|chr| {
            if chr.is_alphanumeric() {
                format!("{}+", chr)
            } else {
                chr.into()
            }
        })
        .collect::<String>()
}

async fn apply_rule(
    message: &impl MessageLike,
    rule: MessageFilterRule,
    reasons: Vec<String>,
    executor: &ActionExecutor,
) -> Result<()> {
    let mut action_taken = None;
    let guild_id = message.guild_id().unwrap();
    let author_id = message.author().id();
    let channel_id = message.channel_id();
    let message_id = message.id();

    if rule.get_notify_moderator() {
        action_taken = Some(format!(
            "Message filter found notable message by <@{}> in <#{}>",
            author_id, channel_id
        ));

        tracing::info!(
            "Message filter notified moderator about message {} in channel {}",
            message.id(),
            message.channel_id()
        );
    }

    if rule.get_delete_message() {
        action_taken = Some(format!(
            "Message filter deleted a message by <@{}> in <#{}>",
            author_id, channel_id
        ));

        // Delete the message from the cache to avoid logging it when it gets deleted.
        CachedMessage::delete(message.channel_id(), message.id())
            .query_async(&mut executor.storage().clone())
            .await?;

        let http = executor.http().clone();
        tokio::spawn(async move {
            // TODO(james7132): DM the user that their message was deleted.
            let result = http.delete_message(channel_id, message_id).exec().await;
            if let Err(err) = result {
                tracing::error!(
                    "Error while deleting message {} in channel {} for message filter: {}",
                    message_id,
                    channel_id,
                    err
                );
                return;
            }
        });

        tracing::info!(
            "Message filter deleted message {} in channel {}",
            message.id(),
            message.channel_id()
        );
    }

    if !rule.additional_actions.is_empty() {
        let rule = rule.clone();
        let exec = executor.clone();
        tokio::spawn(async move {
            for action_template in rule.additional_actions.iter() {
                let mut action = action_template.clone();
                action.set_guild_id(guild_id.get());
                action.set_user_id(author_id.get());
                action.set_reason(format!("Triggered message filter: {}", rule.get_name()));
                if let Err(err) = exec.execute_action(&action).await {
                    tracing::error!("Error while running actions for message filter: {}", err);
                    break;
                }
            }
        });
    }

    if let Some(action_taken) = action_taken {
        let ping = if rule.get_notify_moderator() {
            let (_, ping) = hourai_storage::ping_online_mod(guild_id, executor.storage()).await?;
            ping
        } else {
            "".to_string()
        };

        let response = format!(
            "{} {}:\n```\n   - {}\n```",
            ping,
            action_taken,
            reasons.join("\n   - ")
        );

        let config: LoggingConfig =
            GuildConfig::fetch_or_default(guild_id, &mut executor.storage().redis().clone())
                .await?;
        if let Some(modlog_id) = ChannelId::new(config.get_modlog_channel_id()) {
            executor
                .http()
                .create_message(modlog_id)
                .content(&response)?
                .embeds(&[message_logging::message_to_embed(message)?.build()?])?
                .exec()
                .await?;
        }
    }

    Ok(())
}

async fn get_filter_reasons(
    moderator: bool,
    message: &impl MessageLike,
    criteria: &MessageFilterRule_Criteria,
) -> Result<Vec<String>> {
    let mut reasons = Vec::new();

    let is_bot = criteria.get_exclude_bots() && message.author().bot();
    let is_in_excluded_channel = criteria
        .get_excluded_channels()
        .contains(&message.channel_id().get());
    let is_moderator = criteria.get_exclude_moderators() && moderator;

    if is_bot || is_moderator || is_in_excluded_channel {
        return Ok(reasons);
    }

    match RegexSet::new(&criteria.matches) {
        Ok(regex) => {
            if regex.is_match(message.content()) {
                reasons.push(String::from("Message contains banned word or phrase."));
            }
        }
        Err(err) => {
            tracing::warn!(
                "Error while building regex for message filter: {}",
                err
            );
        }
    }

    if criteria.get_includes_slurs() {
        for word in message.content().split_whitespace() {
            if SLUR_REGEX.is_match(word) {
                reasons.push(format!("Message contains recognized racial slur: {}", word));
            }
            break;
        }
    }

    if criteria.get_includes_invite_links() {
        if DISCORD_INVITE_REGEX.is_match(message.content()) {
            reasons.push("Message contains Discord invite link.".into());
        }
    }

    if let Some(mentions) = criteria.mentions.as_ref() {
        get_mention_reason(message, mentions, &mut reasons);
    }
    if let Some(embeds) = criteria.embeds.as_ref() {
        get_embed_reason(message, embeds, &mut reasons);
    }

    Ok(reasons)
}

fn get_mention_reason(
    message: &impl MessageLike,
    criteria: &MentionFilterCriteria,
    reasons: &mut Vec<String>,
) {
    let users = mentions::get_user_mention_ids(message.content())
        .map(|id| id.get())
        .collect::<Vec<_>>();
    let roles = mentions::get_role_mention_ids(message.content())
        .map(|id| id.get())
        .collect::<Vec<_>>();
    let mut all = Vec::new();
    all.extend(users.iter().cloned());
    all.extend(roles.iter().cloned());

    check_limits("user mentions", users, criteria.get_user_mention(), reasons);
    check_limits("role mentions", roles, criteria.get_role_mention(), reasons);
    check_limits("mentions", all, criteria.get_any_mention(), reasons);
}

fn get_embed_reason(
    message: &impl MessageLike,
    criteria: &EmbedFilterCriteria,
    reasons: &mut Vec<String>,
) {
    let mut urls = HashSet::new();
    urls.extend(
        message
            .embeds()
            .iter()
            .filter_map(|embed| embed.url.clone()),
    );
    urls.extend(message.attachments().iter().map(|embed| embed.url.clone()));
    if criteria.has_max_embed_count() && urls.len() > criteria.get_max_embed_count() as usize {
        reasons.push(format!(
            "Message has {} embeds or attachments. More than the server maximum of {}.",
            urls.len(),
            criteria.get_max_embed_count()
        ));
    }
}

fn check_limits(
    name: &str,
    ids: Vec<u64>,
    limits: &MentionFilterCriteria_MentionLimits,
    reasons: &mut Vec<String>,
) {
    let unique = ids.iter().cloned().collect::<HashSet<_>>();
    if limits.has_maximum_total() {
        if ids.len() > limits.get_maximum_total() as usize {
            reasons.push(format!(
                "Total {} more than the server limit (seen: {}, limit: {}).",
                name,
                ids.len(),
                limits.get_maximum_total()
            ));
        }
    }

    if limits.has_maximum_unique() {
        if unique.len() > limits.get_maximum_unique() as usize {
            reasons.push(format!(
                "Unique {} more than the server limit (seen: {}, limit: {}).",
                name,
                ids.len(),
                limits.get_maximum_unique()
            ));
        }
    }
}
